lib/archive/unpack_unix.go

```diff
@@ -202,6 +202,56 @@ func streamCopy(op trace.Operation, stdin io.WriteCloser, tarStream io.Reader) e
 	}
 }
 
+func DockerUnpack(op trace.Operation, root string, tarStream io.Reader) (int64, error) {
+	fi, err := os.Stat(root)
+	if err != nil {
+		// the target unpack path does not exist. We should not get here.
+		op.Errorf("tar unpack target does not exist: %s", root)
+		return 0, err
+	}
+
+	if !fi.IsDir() {
+		err := fmt.Errorf("unpack root target is not a directory: %s", root)
+		op.Error(err)
+		return 0, err
+	}
+
+	// #nosec: 193 applianceBinaryPath is a constant, not a variable
+	cmd := exec.Command(string(applianceBinaryPath), op.ID(), root)
+
+	stdin, err := cmd.StdinPipe()
+
+	if err != nil {
+		op.Error(err)
+		return 0, err
+	}
+
+	if stdin == nil {
+		err = errors.New("stdin was nil")
+		op.Error(err)
+		return 0, err
+	}
+
+	var n int64
+	go func(n int64) {
+
+		defer stdin.Close()
+		if n, err = io.Copy(stdin, tarStream); err != nil {
+			op.Errorf("Error copying tarStream: %s", err.Error())
+		}
+
+	}(n)
+
+	if err = cmd.Start(); err != nil {
```

Comment:
You are starting a goroutine before you call start. Does that goroutine complete its work if cmd.Start fails?

Reply:
no but it blocks while waiting to read in `io.Copy`

Reply:
no, this should go after start and before wait, i realize. it also doesn't complete its work if EOF is never seen, same as w/ `unpack`.. @hickeng and I were trying to think of a way to avoid that problem, but weren't able to think of anything sufficiently reliable and uncomplicated. If you have insight into that @vburenin it would be much appreciated feedback.